<a href="https://colab.research.google.com/github/everythingcanbedone/everythingcanbedone/blob/main/Go_digital_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
insurance_data=pd.read_csv('insurance_data.csv')
employee_data=pd.read_csv('employee_data.csv')
vendor_data=pd.read_csv('vendor_data.csv')

Task # 1

Merge the 3 dataset and create 1 view of data.
You can merge insurance_data.csv and employee_data.csv on
AGENT_ID You can merge insurance_data.csv and vendor_data.csv on
VENDOR_ID Note: Use left Outer join as not all claims require Vendor

In [ ]:
output=pd.merge(insurance_data,employee_data,on='AGENT_ID',how='left',suffixes=('_CUSTOMER','_AGENT'))
Task1=pd.merge(output,vendor_data,on='VENDOR_ID',how='left')

In [ ]:
Task1

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER  \
0     2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468   
1     2020-06-01 00:00:00    TXN00000002   A00008149   PLC00009594   
2     2020-06-01 00:00:00    TXN00000003   A00003172   PLC00007969   
3     2020-06-01 00:00:00    TXN00000004   A00007572   PLC00009292   
4     2020-06-01 00:00:00    TXN00000005   A00008173   PLC00000204   
...                   ...            ...         ...           ...   
9995  2021-06-30 00:00:00    TXN00009996   A00006029   PLC00007245   
9996  2021-06-30 00:00:00    TXN00009997   A00003555   PLC00002590   
9997  2021-06-30 00:00:00    TXN00009998   A00004230   PLC00006613   
9998  2021-06-30 00:00:00    TXN00009999   A00006694   PLC00001781   
9999  2021-06-30 00:00:00    TXN00010000   A00000390   PLC00005197   

     POLICY_EFF_DT     LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  \
0       2015-06-23  2020-05-16  2020-05-21         Health          157.13   
1       2018-04-21  2020-05-13  2020-05-18       Property          141.71   
2       2019-10-03  2020-05-21  2020-05-26       Property          157.24   
3       2016-11-29  2020-05-14  2020-05-19         Health          172.87   
4       2011-12-26  2020-05-17  2020-05-22         Travel           88.53   
...            ...         ...         ...            ...             ...   
9995    2014-04-11  2021-06-25  2021-06-26          Motor          105.75   
9996    2015-08-25  2021-06-12  2021-06-17           Life           64.59   
9997    2018-10-11  2021-06-18  2021-06-23         Health          183.50   
9998    2016-01-29  2021-06-16  2021-06-21          Motor          113.46   
9999    2014-09-17  2021-06-21  2021-06-24         Mobile            8.22   

      CLAIM_AMOUNT  ... STATE_AGENT POSTAL_CODE_AGENT EMP_ROUTING_NUMBER  \
0             9000  ...          TN             37217           35441155   
1            26000  ...          AZ             85301           44055451   
2            13000  ...          AL             36105           55770246   
3            16000  ...          OK             73071           87652976   
4             3000  ...          MD             21061           76119019   
...            ...  ...         ...               ...                ...   
9995          1000  ...          AR             72703           35413882   
9996         67000  ...          GA             31405           91206790   
9997          8000  ...          KY             40272          115241097   
9998          2000  ...          AR             72703           13995380   
9999           300  ...          OK             73003           39790996   

         EMP_ACCT_NUMBER                       VENDOR_NAME  \
0     LRZO53254369804934                        Ramsey Inc   
1     XBYG14382904542874  Morales, Campbell and Washington   
2     WRXH14597428306333                       Johnson Ltd   
3     YEPN16077902472499                  Blackwell-Foster   
4     HZBK86544578578547                       Evans-Scott   
...                  ...                               ...   
9995  LFOJ76068590051807                               NaN   
9996  AHEF60771774635030          Newton, Hodges and Henry   
9997  WREH18805592939657                    Reese-Mckinney   
9998  AAZI21883736448048       Harmon, Manning and Bradley   
9999  RPES05613584687957                    Foster-Sanchez   

                  ADDRESS_LINE1 ADDRESS_LINE2          CITY  STATE  \
0          3114 West 20th Court           NaN   Panama City     FL   
1         42 West Louise Street             B  Fayetteville     AR   
2       2619 North Quality Lane          #315  Fayetteville     AR   
3             3609 Meadow Drive           NaN     Nashville     TN   
4     4001 19th Place Northeast           NaN    Washington     DC   
...                         ...           ...           ...    ...   
9995                        NaN           NaN           NaN    NaN   
9996      32512 Christine Dr

Task #2


Business Leader wants to find Top 3 Insurance Type where we are getting
most insurance claims?

In [ ]:
insurance_data.INSURANCE_TYPE.value_counts()[:3]

Property    1692
Mobile      1692
Health      1690
Name: INSURANCE_TYPE, dtype: int64

Top 3 Insurance Type which are getting most insurance claims are

Mobile with 1692 claims

Property with 1692 claims

Heath with 1690 claims

Task #3

Business Leader wants to find Top 5 States where we are getting most insurance claims
for customer belonging to HIGH(H) risk segment?

In [ ]:
data = insurance_data['RISK_SEGMENTATION'] == "H"
insurance_data[data]["STATE"].value_counts()[:5]

CA    148
AZ    105
FL    104
TN    100
AR     97
Name: STATE, dtype: int64

Top 5 States where we are getting most insurance claims for customers belonging to HIGH(H) risk segment are

    CA with 148 claims
    AZ with 105 claims
    FL with 104 claims
    TN with 100 claims
    AR with 97 claims



Task #4

Business wants to create a new variable “COLOCATION” which will have following values IF
Customer State == Incident State == Agent Address State THEN 1 ELSE 0
Find the mean of this new column

In [ ]:
Task1.columns

Index(['TXN_DATE_TIME', 'TRANSACTION_ID', 'CUSTOMER_ID', 'POLICY_NUMBER',
       'POLICY_EFF_DT', 'LOSS_DT', 'REPORT_DT', 'INSURANCE_TYPE',
       'PREMIUM_AMOUNT', 'CLAIM_AMOUNT', 'CUSTOMER_NAME',
       'ADDRESS_LINE1_CUSTOMER', 'ADDRESS_LINE2_CUSTOMER', 'CITY_CUSTOMER',
       'STATE_CUSTOMER', 'POSTAL_CODE_CUSTOMER', 'SSN', 'MARITAL_STATUS',
       'AGE', 'TENURE', 'EMPLOYMENT_STATUS', 'NO_OF_FAMILY_MEMBERS',
       'RISK_SEGMENTATION', 'HOUSE_TYPE', 'SOCIAL_CLASS', 'ROUTING_NUMBER',
       'ACCT_NUMBER', 'CUSTOMER_EDUCATION_LEVEL', 'CLAIM_STATUS',
       'INCIDENT_SEVERITY', 'AUTHORITY_CONTACTED', 'ANY_INJURY',
       'POLICE_REPORT_AVAILABLE', 'INCIDENT_STATE', 'INCIDENT_CITY',
       'INCIDENT_HOUR_OF_THE_DAY', 'AGENT_ID', 'VENDOR_ID', 'AGENT_NAME',
       'DATE_OF_JOINING', 'ADDRESS_LINE1_AGENT', 'ADDRESS_LINE2_AGENT',
       'CITY_AGENT', 'STATE_AGENT', 'POSTAL_CODE_AGENT', 'EMP_ROUTING_NUMBER',
       'EMP_ACCT_NUMBER', 'VENDOR_NAME', 'ADDRESS_LINE1', 'ADDRESS_LINE2',
       

In [ ]:
def categorise(Task1):  
    if ((Task1['INCIDENT_STATE']) == (Task1["STATE_AGENT"]))  and  ((Task1["STATE_AGENT"]) == (Task1["STATE_CUSTOMER"])):
      val=1
    else:
      val=  0
    return val

In [ ]:
Task1['COLOCATION'] = Task1.apply(categorise, axis=1)

In [ ]:
Task1["COLOCATION"].value_counts()

0    9956
1      44
Name: COLOCATION, dtype: int64

In [ ]:
Task1['COLOCATION'].mean()

0.0044

Mean of Colocation column is 00.44

Task #5

Data entry error was detected in the data and you are required to correct it. If for any
claim transaction “AUTHORITY_CONTACTED” is NOT “Police” and POLICE_AVAILABLE == 1
Then Update “AUTHORITY_CONTACTED” to Police.


In [ ]:
Task1.columns

Index(['TXN_DATE_TIME', 'TRANSACTION_ID', 'CUSTOMER_ID', 'POLICY_NUMBER',
       'POLICY_EFF_DT', 'LOSS_DT', 'REPORT_DT', 'INSURANCE_TYPE',
       'PREMIUM_AMOUNT', 'CLAIM_AMOUNT', 'CUSTOMER_NAME',
       'ADDRESS_LINE1_CUSTOMER', 'ADDRESS_LINE2_CUSTOMER', 'CITY_CUSTOMER',
       'STATE_CUSTOMER', 'POSTAL_CODE_CUSTOMER', 'SSN', 'MARITAL_STATUS',
       'AGE', 'TENURE', 'EMPLOYMENT_STATUS', 'NO_OF_FAMILY_MEMBERS',
       'RISK_SEGMENTATION', 'HOUSE_TYPE', 'SOCIAL_CLASS', 'ROUTING_NUMBER',
       'ACCT_NUMBER', 'CUSTOMER_EDUCATION_LEVEL', 'CLAIM_STATUS',
       'INCIDENT_SEVERITY', 'AUTHORITY_CONTACTED', 'ANY_INJURY',
       'POLICE_REPORT_AVAILABLE', 'INCIDENT_STATE', 'INCIDENT_CITY',
       'INCIDENT_HOUR_OF_THE_DAY', 'AGENT_ID', 'VENDOR_ID', 'AGENT_NAME',
       'DATE_OF_JOINING', 'ADDRESS_LINE1_AGENT', 'ADDRESS_LINE2_AGENT',
       'CITY_AGENT', 'STATE_AGENT', 'POSTAL_CODE_AGENT', 'EMP_ROUTING_NUMBER',
       'EMP_ACCT_NUMBER', 'VENDOR_NAME', 'ADDRESS_LINE1', 'ADDRESS_LINE2',
       

In [ ]:
Task1['AUTHORITY_CONTACTED'] = np.where(
   (Task1['AUTHORITY_CONTACTED'] !="POLICE") & (Task1['POLICE_REPORT_AVAILABLE'] ==1) ,"Police", Task1['AUTHORITY_CONTACTED']
   )


In [ ]:
Task1[(Task1['POLICE_REPORT_AVAILABLE']==1)&(Task1['AUTHORITY_CONTACTED']!='Police')]

Empty DataFrame
Columns: [TXN_DATE_TIME, TRANSACTION_ID, CUSTOMER_ID, POLICY_NUMBER, POLICY_EFF_DT, LOSS_DT, REPORT_DT, INSURANCE_TYPE, PREMIUM_AMOUNT, CLAIM_AMOUNT, CUSTOMER_NAME, ADDRESS_LINE1_CUSTOMER, ADDRESS_LINE2_CUSTOMER, CITY_CUSTOMER, STATE_CUSTOMER, POSTAL_CODE_CUSTOMER, SSN, MARITAL_STATUS, AGE, TENURE, EMPLOYMENT_STATUS, NO_OF_FAMILY_MEMBERS, RISK_SEGMENTATION, HOUSE_TYPE, SOCIAL_CLASS, ROUTING_NUMBER, ACCT_NUMBER, CUSTOMER_EDUCATION_LEVEL, CLAIM_STATUS, INCIDENT_SEVERITY, AUTHORITY_CONTACTED, ANY_INJURY, POLICE_REPORT_AVAILABLE, INCIDENT_STATE, INCIDENT_CITY, INCIDENT_HOUR_OF_THE_DAY, AGENT_ID, VENDOR_ID, AGENT_NAME, DATE_OF_JOINING, ADDRESS_LINE1_AGENT, ADDRESS_LINE2_AGENT, CITY_AGENT, STATE_AGENT, POSTAL_CODE_AGENT, EMP_ROUTING_NUMBER, EMP_ACCT_NUMBER, VENDOR_NAME, ADDRESS_LINE1, ADDRESS_LINE2, CITY, STATE, POSTAL_CODE, COLOCATION]
Index: []

[0 rows x 54 columns]

In [ ]:
Task1["AUTHORITY_CONTACTED"].value_counts()

Police       7852
Ambulance    1261
None          578
Other         309
Name: AUTHORITY_CONTACTED, dtype: int64

Task #6

Business wants to check the Claim Amount for deviation for each transaction, they would
like you to calculate as follow
CLAIM_DEVIATION = AVG_CLAIM_AMOUNT_FOR_LAST_30DAYS (same insurance type)
/ CURRENT_CLAIM_AMOUNT
If the value < 0.5 THEN CLAIM_DEVIATION = 1 ELSE 0
If there is less than 30 days of transaction history THEN
-1 Note: LAST_30DAYS does not include current day

In [ ]:
Task1.TXN_DATE_TIME.max()

'2021-06-30 00:00:00'

In [ ]:

curr_date = pd.to_datetime('2021-06-30')
Task1['TXN_DATE_TIME'] = pd.to_datetime(Task1['TXN_DATE_TIME'])
last_30 = Task1.loc[(Task1['TXN_DATE_TIME'] > '2021-05-30') & (Task1['TXN_DATE_TIME'] < '2021-06-30')]  

In [ ]:
average_claim = last_30.groupby('INSURANCE_TYPE').mean()['CLAIM_AMOUNT'].to_dict()

In [ ]:
# defining a function for calculating claim_deviation
def Claim_deviation(row):
    if (curr_date - row['TXN_DATE_TIME'])<pd.Timedelta(days=30):
        return -1
    claim_deviation = average_claim[row['INSURANCE_TYPE']] / row['CLAIM_AMOUNT']
    if claim_deviation < 0.5:
        return 1
    else:
        return 0

In [ ]:
Task1['CLAIM_DEVIATION'] = Task1.apply(Claim_deviation,axis=1)

In [ ]:
Task1.CLAIM_DEVIATION.value_counts()

 0    9216
-1     784
Name: CLAIM_DEVIATION, dtype: int64

There are 9216 claims with deviation less than 0.5
784 claims with transaction history of less than 30 days


Task #7


Find All Agents who have worked on more than 2 types of Insurance Claims. Sort them by
Total Claim Amount Approved under them in descending order

In [ ]:
type_count = Task1.groupby('AGENT_ID')['INSURANCE_TYPE'].nunique()

In [ ]:
agent_insurance_types = type_count>2
multi_type_agents = agent_insurance_types[agent_insurance_types].index
Top_agents = Task1.groupby(['AGENT_ID','AGENT_NAME']).sum().loc[multi_type_agents][['CLAIM_AMOUNT']].sort_values('CLAIM_AMOUNT',ascending=False)
Top_agents.join(type_count)

CLAIM_AMOUNT  INSURANCE_TYPE
AGENT_ID   AGENT_NAME                                    
AGENT00807 Don Filkins             528800               6
AGENT00679 Clara Barnett           489000               4
AGENT00771 Roger Burns             422100               5
AGENT00125 Salvador Soriano        400400               4
AGENT00789 Alison Baron            392900               4
...                                   ...             ...
AGENT00910 Vincent Ragan            15500               3
AGENT00689 Connie Hunt              13700               3
AGENT00732 Sylvia Tran              11300               3
AGENT00706 Laura Staggs             10800               3
AGENT00604 Thelma Salinas            5500               3

[1164 rows x 2 columns]

Task #8


Mobile & Travel Insurance premium are discounted by 10%
Health and Property Insurance premium are increased by 7%
Life and Motor Insurance premium are marginally increased by 2%
What will be overall change in % of the Premium Amount Collected for all these Customer?

In [ ]:
premium_collected_past = Task1['PREMIUM_AMOUNT'].sum()
prem_types = view.groupby('INSURANCE_TYPE').sum()['PREMIUM_AMOUNT'].to_dict()
prem_types['Mobile']=prem_types['Mobile']*0.9
prem_types['Travel']=prem_types['Travel']*0.9
prem_types['Health']=prem_types['Health']*1.07
prem_types['Property']=prem_types['Property']*1.07
prem_types['Life']=prem_types['Life']*1.02
prem_types['Motor']=prem_types['Motor']*1.02
revised_premium = sum(prem_types.values())
(revised_premium-premium_collected_past)*100/premium_collected_past

-27.07189348107943

In [ ]:
premium_collected_past,revised_premium

(885085.95, 645476.4244)


# New SectionOverall change in Premium amount collected is approximately 2.68% 


# Task #9

Business wants to give discount to customer who are loyal and under stress due to Covid
19. They have laid down an eligibility Criteria as follow
IF CUSTOMER_TENURE > 60 AND EMPLOYMENT_STATUS = “N”
AND NO_OF_FAMILY_MEMBERS >=4 THEN 1 ELSE 0

Create a new column “ELIGIBLE_FOR_DISCOUNT” and find it
mean.

In [ ]:
discount=(Task1['TENURE']>60) & (Task1['EMPLOYMENT_STATUS']=='N')&(Task1['NO_OF_FAMILY_MEMBERS']>=4)
Task1['ELIGIBLE_FOR_DISCOUNT'] = discount.astype('int')
Task1.ELIGIBLE_FOR_DISCOUNT.mean()

0.0299

Mean is 0.0299

# Task #10
Business wants to check Claim Velocity which is defined as follow
CLAIM_VELOCITY = NO_OF_CLAIMS_IN_LAST30DAYS (for the current insurance type)
/ NO_OF_CLAIMS_IN_LAST3DAYS (for the current insurance type)
Note: LAST30DAYS & LAST3DAYS does not include current
day

In [ ]:
view.TXN_DATE_TIME.max()

'2021-03-09 00:00:00'

In [ ]:

curr_date = pd.to_datetime('2021-06-30')
view['CLAIM_STATUS'].value_counts()


A    6777
D     357
Name: CLAIM_STATUS, dtype: int64

# Task #11
Find all low performing agents i.e. employees who are in the bottom 5 percentile based
on Claims worked by them.

In [ ]:
bottom_limit = Task1.groupby('AGENT_ID').sum().quantile(0.05)['CLAIM_AMOUNT']
Agent_data = Task1.groupby('AGENT_ID').sum()

Low_performing_agents = Agent_data[Agent_data['CLAIM_AMOUNT']<=bottom_limit][['CLAIM_AMOUNT']]

In [ ]:
Low_performing_agents

CLAIM_AMOUNT
AGENT_ID                
AGENT00006         22400
AGENT00014          7400
AGENT00016         27000
AGENT00085         25700
AGENT00088         27700
AGENT00091         30000
AGENT00098         28200
AGENT00134         22000
AGENT00149         28900
AGENT00156         13200
AGENT00172         22200
AGENT00185         16100
AGENT00186         27700
AGENT00188         28400
AGENT00197         18500
AGENT00252         20000
AGENT00263         11500
AGENT00275         16200
AGENT00279         28000
AGENT00300         12600
AGENT00310         28800
AGENT00316         22200
AGENT00351         21700
AGENT00425         17800
AGENT00442         30300
AGENT00536         20000
AGENT00562         21100
AGENT00571         24200
AGENT00585         16600
AGENT00590         22200
AGENT00604          5500
AGENT00621          5300
AGENT00628          6100
AGENT00672         21300
AGENT00689         13700
AGENT00706         10800
AGENT00721          9000
AGENT00732         11300
AGENT00773         15000
AGENT00780         11300
AGENT00790          6000
AGENT00804         20000
AGENT00805         22600
AGENT00830         23500
AGENT00841         17800
AGENT00856         24200
AGENT00894         27600
AGENT00910         15500
AGENT00969         28300
AGENT00985         30000
AGENT01021         24300
AGENT01037          5600
AGENT01052         28000
AGENT01077          4200
AGENT01107         24700
AGENT01128         28000
AGENT01131         27500
AGENT01151         28000
AGENT01154           800
AGENT01179         27300

# Task #12
Business wants to find all Suspicious Employees (Agents).
IF TOTAL CLAIM AMOUNT which meet below criteria is >= 15000 THEN AGENT is classified
as Suspicious ELSE Not
CLAIM_STATUS = Approved AND CUSTOMER_RISK_SEGMENTATION = High
AND INCIDENT_SEVERITY = “Major Loss”
If Suspicious, then 1 ELSE 0. Find mean of this column.

In [ ]:
criteria = (Task1['CLAIM_STATUS']=='A')&(Task1['RISK_SEGMENTATION']=='H')&(Task1['INCIDENT_SEVERITY']=='Major Loss')
criteriamet = view[criteria]
grouped_suspicion_data = criteriamet.groupby('AGENT_ID').sum()
suspicious_agents = list(grouped_suspicion_data[grouped_suspicion_data['CLAIM_AMOUNT']>=15000].index)
employee_data['SUSPICIOUS'] = 0
sus_emp_slice = employee_data['AGENT_ID'].isin(suspicious_agents)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


In [ ]:
employee_data.loc[sus_emp_slice,'SUSPICIOUS']=1
employee_data['SUSPICIOUS'].mean()

0.08666666666666667

The mean of suspicious column is 0.124167 that means from all employees 12.42% are suspicious